src/combinators/mod.rs

```diff
@@ -1,42 +1,96 @@
 use crate::gen::GenError;
-use crate::lib::std::io::{self, SeekFrom, Write};
+use crate::lib::std::io::{self, SeekFrom, Write, Seek as _};
 
 /// Holds the result of serializing functions
 ///
 /// The `Ok` case returns the `Write` used for writing, in the `Err` case an instance of
 /// `cookie_factory::GenError` is returned.
-pub type GenResult<I> = Result<I, GenError>;
+pub type GenResult<I> = Result<WriteContext<I>, GenError>;
 
 /// Trait for serializing functions
 ///
 /// Serializing functions take one input `I` that is the target of writing and return an instance
 /// of `cookie_factory::GenResult`.
 ///
 /// This trait is implemented for all `Fn(I) -> GenResult<I>`.
-pub trait SerializeFn<I>: Fn(I) -> GenResult<I> {}
+pub trait SerializeFn<I>: Fn(WriteContext<I>) -> GenResult<I> {}
 
-impl<I, F: Fn(I) -> GenResult<I>> SerializeFn<I> for F {}
+impl<I, F: Fn(WriteContext<I>) -> GenResult<I>> SerializeFn<I> for F {}
+
+pub struct WriteContext<W> {
+    pub write: W,
+    pub position: u64,
+}
+
+impl<W: Write> From<W> for WriteContext<W> {
+    fn from(write: W) -> Self {
+        Self {
+            write,
+            position: 0,
+        }
+    }
+}
+
+impl<W: Write> WriteContext<W> {
+    pub fn into_inner(self) -> (W, u64) {
+        (self.write, self.position)
+    }
+}
+
+impl<W: Write> Write for WriteContext<W> {
+    fn write(&mut self, data: &[u8]) -> crate::lib::std::io::Result<usize> {
+        let amt = self.write.write(data)?;
+        self.position += amt as u64;
+        Ok(amt)
+    }
+
+    #[cfg(feature = "std")]
+    fn flush(&mut self) -> io::Result<()> {
+        self.write.flush()
+    }
+}
+
+impl<W: Seek> io::Seek for WriteContext<W> {
+    fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
+        let old_pos = self.write.seek(SeekFrom::Current(0))?;
+        let new_pos = self.write.seek(pos)?;
+        if new_pos >= old_pos {
+            self.position += new_pos - old_pos;
+        } else {
+            self.position -= old_pos - new_pos;
+        }
+        Ok(new_pos)
+    }
+}
+
+pub fn gen<W: Write, F: SerializeFn<W>>(f: F, w: W) -> GenResult<W> {
```

Comment:
Should this return a `GenResult` or a `Result<(W, u64), GenError>`?

Reply:
From updating the tests/docs, I think returning the unwrapped `Result<(W, u64), GenError>` is more ergonomic.

Reply:
@Keruspe Let me know what you think, then I'll continue finishing this :)

Reply:
Fine with returning the unwrapped data.
Is it valid/would it make sense to just `type WriteContext<W> = (W, u64);`?

Reply:
Then we can't implement `from()` (so you always have to init with 0) or `into_inner()` :) I'd keep it an actual struct so we can also add other things there if needed.

Do you want to call it `WriteContext` or `Writer`?

Reply:
Both are fine by me